common/auth/service_principal_key_provider.go

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2016, 2018, Oracle and/or its affiliates. All rights reserved.
+
+package auth
+
+import (
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/oracle/oci-go-sdk/common"
+)
+
+type servicePrincipalKeyProvider struct {
+	federationClient federationClient
+}
+
+func newServicePrincipalKeyProvider(tenancyID, region, cert, key string, intermediates []string, passphrase *string) (provider *servicePrincipalKeyProvider, err error) {
+	leafCertificateRetriever := newStaticX509CertificateRetriever([]byte(cert), []byte(key), passphrase)
+
+	intermediateCertificateRetrievers := []x509CertificateRetriever{}
+	for _, intermediate := range intermediates {
+		intermediateCertificateRetrievers =
+			append(intermediateCertificateRetrievers, newStaticX509CertificateRetriever([]byte(intermediate), []byte(key), passphrase))
+	}
+
+	federationClient := newX509FederationClient(
+		common.Region(region), tenancyID, leafCertificateRetriever, intermediateCertificateRetrievers, false)
+
+	provider = &servicePrincipalKeyProvider{federationClient: federationClient}
+	return
+}
+
+func (p *servicePrincipalKeyProvider) PrivateRSAKey() (privateKey *rsa.PrivateKey, err error) {
```

Comment:
I have been wondering about this, I have been asked to support signing without material access to the private key(hw signer). I wonder how easy/hard will be to make all these providers support that.